### PR TITLE
perf: Remove one redundant CopyExec for SMJ

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

We were injecting too many CopyExecs for SMJ. Plans looked like this:

```
SortMergeJoin
  CopyExec [UnpackOrDeepClone]
    SortExec
      CopyExec [UnpackOrDeepClone]
        ...
  CopyExec [UnpackOrDeepClone]
    SortExec
      CopyExec [UnpackOrDeepClone]
        ...
```

## What changes are included in this PR?

With the changes in this PR, the plan now looks like:

```
SortMergeJoin
  SortExec
    CopyExec [UnpackOrDeepClone]
      ...
  SortExec
    CopyExec [UnpackOrDeepClone]
      ...
```

## How are these changes tested?

Existing tests

### CODE REVIEW

File: native/core/src/execution/datafusion/planner.rs [hunk 1]
```diff
@@ -1290,6 +1285,21 @@ impl PhysicalPlanner {
         ))
     }
 
+    fn wrap_in_copy_exec(
+        is_sort_merge: bool,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> Arc<dyn ExecutionPlan> {
+        if is_sort_merge {
```
Comment (mbutrovich) [inline native/core/src/execution/datafusion/planner.rs:1292]: Is there a more general description for SMJ's property to name this variable, and then we could use this function for the duplicated code elsewhere in this file?
https://github.com/apache/datafusion-comet/blob/31d0b0a38165e9315a2c0d981406dd48f7f7a3af/native/core/src/execution/datafusion/planner.rs#L920

**Comment by andygrove** [comment]
I ran TPC-DS benchmarks comparing the latest from the main branch with this PR, and somehow, this PR introduces a considerable regression in performance (~20%). I am investigating.

File: native/core/src/execution/datafusion/planner.rs (hunk 1 above)
Comment (andygrove) [reply native/core/src/execution/datafusion/planner.rs:1292]: Yes, I will refactor this to reduce duplication

**Comment by andygrove** [comment]
A single run of TPC-DS comparing main to this PR does not show a significant difference (~1% difference). I did not expect this to make much difference, but it does remove a redundant operator and implied the plan,

File: native/core/src/execution/datafusion/planner.rs [hunk 2]
```diff
@@ -917,15 +917,11 @@ impl PhysicalPlanner {
 
                 let fetch = sort.fetch.map(|num| num as usize);
 
-                let copy_exec = if can_reuse_input_batch(&child) {
-                    Arc::new(CopyExec::new(child, CopyMode::UnpackOrDeepCopy))
-                } else {
-                    Arc::new(CopyExec::new(child, CopyMode::UnpackOrClone))
-                };
+                let child = Self::wrap_in_copy_exec(child);
```
Comment (mbutrovich) [inline native/core/src/execution/datafusion/planner.rs:920]: Can you add a similar comment here as you added with HashJoin on why sort needs a CopyExec?

**Review by mbutrovich — APPROVED** [review]
This a great catch, and a super clean refactor.

**Review by kazuyukitanimura — APPROVED** [review]

**Pull request merged**